File: go/euler041.go
// https://projecteuler.net/problem=41

/*

Pandigital Prime

We shall say that an n-digit number is pandigital if it makes use of all the digits 1 to n exactly once. For example, 2143 is a 4-digit pandigital and is also prime.

What is the largest n-digit pandigital prime that exists?

Answer: 7652413

*/

package main

import "fmt"

func isPrime(n int) bool {
	if n <= 1 {
		return false
	}
	if n <= 3 {
		return true
	}
	if n%2 == 0 || n%3 == 0 {
		return false
	}
	for i := 5; i*i <= n; i += 6 {
		if n%i == 0 || n%(i+2) == 0 {
			return false
		}
	}
	return true
}

func permutations(arr []int, start int, result *[]int) {
	if start == len(arr)-1 {
		num := 0
		for _, d := range arr {
			num = num*10 + d
		}
		*result = append(*result, num)
		return
	}
	for i := start; i < len(arr); i++ {
		arr[start], arr[i] = arr[i], arr[start]
		permutations(arr, start+1, result)
		arr[start], arr[i] = arr[i], arr[start]
	}
}

func main() {
	maxPrime := 0
	for n := 7; n >= 1; n-- {
		digits := make([]int, n)
		for i := 0; i < n; i++ {
			digits[i] = n - i
		}
		var nums []int
		permutations(digits, 0, &nums)
		for _, num := range nums {
			if isPrime(num) && num > maxPrime {
				maxPrime = num
			}
		}
		if maxPrime > 0 {
			break
		}
	}
	fmt.Println(maxPrime)
}

File: rust/src/euler041.rs
// https://projecteuler.net/problem=41
//
// We shall say that an n-digit number is pandigital if it makes use of all the digits 1 to n exactly once. For example, 2143 is a 4-digit pandigital and is also prime.
//
// What is the largest n-digit pandigital prime that exists?
//
// Answer: 7652413

pub fn pandigital_prime() -> u64 {
    // n=7 is the largest possible, since n=8 and n=9 sums are divisible by 3
    let digits = [1,2,3,4,5,6,7];
        let mut max_prime = 0u64;
    for perm in permutations(&digits) {
        let num = to_number(&perm);
        if is_prime(num) && num > max_prime {
            max_prime = num;
        }
    }
    max_prime
}

fn is_prime(n: u64) -> bool {
    if n < 2 { return false; }
    if n == 2 || n == 3 { return true; }
    if n % 2 == 0 || n % 3 == 0 { return false; }
    let mut i = 5;
    while i * i <= n {
        if n % i == 0 || n % (i + 2) == 0 { return false; }
        i += 6;
    }
    true
}

fn permutations(arr: &[u8]) -> Vec<Vec<u8>> {
    let mut result = Vec::new();
    let mut arr = arr.to_vec();
    let len = arr.len();
    heap_permute(&mut arr, len, &mut result);
    result
}

fn heap_permute(arr: &mut [u8], n: usize, result: &mut Vec<Vec<u8>>) {
    if n == 1 {
        result.push(arr.to_vec());
        return;
    }
    for i in 0..n {
        heap_permute(arr, n - 1, result);
        if n % 2 == 1 {
            arr.swap(0, n - 1);
        } else {
            arr.swap(i, n - 1);
        }
    }
}

fn to_number(digits: &[u8]) -> u64 {
    let mut num = 0u64;
    for &d in digits {
        num = num * 10 + d as u64;
    }
    num
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn euler_041() {
        assert_eq!(pandigital_prime(), 7652413);
    }
}